lib/portage/dbapi/bintree.py

```diff
@@ -1203,6 +1302,21 @@ def inject(self, cpv, filename=None):
 				noiselevel=-1)
 			return
 		metadata = self._read_metadata(full_path, s)
+
+		if full_path.endswith(SUPPORTED_XPAK_EXTENSIONS):
+			metadata["BINPKG_FORMAT"] = "xpak"
+		elif full_path.endswith(SUPPORTED_GPKG_EXTENSIONS):
+			metadata["BINPKG_FORMAT"] = "gpkg"
+		else:
+			# check if extra temporary extension exists
+			_filename = os.path.splitext(full_path)[0]
+			if _filename.endswith(SUPPORTED_XPAK_EXTENSIONS):
+				metadata["BINPKG_FORMAT"] = "xpak"
+			elif _filename.endswith(SUPPORTED_GPKG_EXTENSIONS):
+				metadata["BINPKG_FORMAT"] = "gpkg"
+			else:
+				raise InvalidBinaryPackageFormat(full_path)
+
```

Comment:
It will be much nicer if we get BINPKG_FORMAT from the _read_metadata method, since then the temporary file name will be irrelevant. We can assign BINPKG_FORMAT=xpak if `portage.xpak.tbz2(filename).get_data()` returns a non-empty dict, and otherwise assign BINPKG_FORMAT=gpkg if it's a tar file containing a gpkg-1 file.

Reply:
I can't do that because BINPKG_FORMAT is not in the package metadata, which I need know the format before read it.
Maybe I should write a detection function based on the file header.

Reply:
Yes, some kind of detection function is exactly what's needed.

Reply:
Thanks for addressing this in 317abe73191737e1d0d4161b4c3531e434a4eaa2.